lib/naxum/src/generic/handler.rs

```diff
@@ -0,0 +1,235 @@
+where
+    R: Send,
+{
+    type Future: Future<Output = Response> + Send + 'static;
+
+    fn call(self, req: R, state: S) -> Self::Future;
+
+    fn layer<L>(self, layer: L) -> Layered<L, Self, T, S, R>
+    where
+        L: Layer<HandlerService<Self, T, S, R>> + Clone,
+        L::Service: Service<R>,
+    {
+        Layered {
+            layer,
+            handler: self,
+            _marker: PhantomData,
+            _request_marker: PhantomData,
+        }
+    }
+
+    fn with_state(self, state: S) -> HandlerService<Self, T, S, R> {
+        HandlerService::new(self, state)
+    }
+}
+
+impl<F, Fut, Res, S, R> Handler<((),), S, R> for F
+where
+    F: FnOnce() -> Fut + Clone + Send + 'static,
+    Fut: Future<Output = Res> + Send,
+    Res: IntoResponse,
+    R: Send,
+{
+    type Future = Pin<Box<dyn Future<Output = Response> + Send>>;
+
+    fn call(self, _req: R, _state: S) -> Self::Future {
+        Box::pin(async move { self().await.into_response() })
+    }
+}
+
+// FIXME(nick): this is where the pain begins. You need some of the "Message" and "MessageHead" goodies to make this work well.
+// macro_rules! impl_handler {
+//     (
+//         [$($ty:ident),*], $last:ident
+//     ) => {
+//         #[allow(non_snake_case, unused_mut)]
+//         impl<F, Fut, S, R, Res, M, $($ty,)* $last> Handler<(M, $($ty,)* $last,), S, R> for F
+//         where
+//             F: FnOnce($($ty,)* $last,) -> Fut + Clone + Send + 'static,
+//             Fut: Future<Output = Res> + Send,
+//             S: Send + Sync + 'static,
+//             R: Send + 'static,
+//             Res: IntoResponse,
+//             $( $ty: FromMessageHead<S> + Send, )*
+//             $last: FromMessage<S, R, M> + Send,
+//         {
+//             type Future = Pin<Box<dyn Future<Output = Response> + Send>>;
+
+//             fn call(self, req: R, state: S) -> Self::Future {
+//                 Box::pin(async move {
+//                     let (mut parts, body) = req.into_parts();
+//                     let state = &state;
+
+//                     $(
+//                         let $ty = match $ty::from_message_head(&mut parts, state).await {
+//                             Ok(value) => value,
+//                             Err(rejection) => return rejection.into_response(),
+//                         };
+//                     )*
+
+//                     let req = match Message::from_parts(parts, body) {
+//                         Ok(value) => value,
+//                         Err(rejection) => return rejection.into_response(),
+//                     };
+
+//                     let $last = match $last::from_message(req, state).await {
+//                         Ok(value) => value,
+//                         Err(rejection) => return rejection.into_response(),
+//                     };
+
+//                     let res = self($($ty,)* $last,).await;
+
+//                     res.into_response()
+//                 })
```

Comment:
This is where the pain is when you lose the `Message<R>` where `R: MessageHead` goodies.